datafusion/core/src/physical_plan/joins/symmetric_hash_join.rs

```diff
@@ -0,0 +1,2640 @@
+/// (ascending/descending) of the probe side. Here, LV denotes the last value on
+/// the probe side.
+///
+/// As a concrete example, consider the following query:
+///
+///   SELECT * FROM left_table, right_table
+///   WHERE
+///     left_key = right_key AND
+///     a > b - 3 AND
+///     a < b + 10
+///
+/// where columns "a" and "b" come from tables "left_table" and "right_table",
+/// respectively. When a new `RecordBatch` arrives at the right side, the
+/// condition a > b - 3 will possibly indicate a prunable range for the left
+/// side. Conversely, when a new `RecordBatch` arrives at the left side, the
+/// condition a < b + 10 will possibly indicate prunability for the right side.
+/// Let’s inspect what happens when a new RecordBatch` arrives at the right
+/// side (i.e. when the left side is the build side):
+///
+///         Build      Probe
+///       +-------+  +-------+
+///       | a | z |  | b | y |
+///       |+--|--+|  |+--|--+|
+///       | 1 | 2 |  | 4 | 3 |
+///       |+--|--+|  |+--|--+|
+///       | 3 | 1 |  | 4 | 3 |
+///       |+--|--+|  |+--|--+|
+///       | 5 | 7 |  | 6 | 1 |
+///       |+--|--+|  |+--|--+|
+///       | 7 | 1 |  | 6 | 3 |
+///       +-------+  +-------+
+///
+/// In this case, the interval representing viable (i.e. joinable) values for
+/// column "a" is [1, ∞], and the interval representing possible future values
+/// for column "b" is [6, ∞]. With these intervals at hand, we next calculate
+/// intervals for the whole filter expression and propagate join constraint by
+/// traversing the expression graph.
+/// ```
+fn calculate_filter_expr_intervals(
+    build_input_buffer: &RecordBatch,
+    build_sorted_filter_expr: &mut SortedFilterExpr,
+    probe_batch: &RecordBatch,
+    probe_sorted_filter_expr: &mut SortedFilterExpr,
+) -> Result<()> {
+    // If either build or probe side has no data, return early:
+    if build_input_buffer.num_rows() == 0 || probe_batch.num_rows() == 0 {
+        return Ok(());
+    }
+    // Evaluate build side filter expression and convert the result to an array
+    let build_array = build_sorted_filter_expr
+        .origin_sorted_expr()
+        .expr
+        .evaluate(&build_input_buffer.slice(0, 1))?
+        .into_array(1);
+    // Evaluate probe side filter expression and convert the result to an array
+    let probe_array = probe_sorted_filter_expr
+        .origin_sorted_expr()
+        .expr
+        .evaluate(&probe_batch.slice(probe_batch.num_rows() - 1, 1))?
+        .into_array(1);
+
+    // Update intervals for both build and probe side filter expressions
+    for (array, sorted_expr) in vec![
+        (build_array, build_sorted_filter_expr),
+        (probe_array, probe_sorted_filter_expr),
+    ] {
+        // Convert the array to a ScalarValue:
+        let value = ScalarValue::try_from_array(&array, 0)?;
+        // Create a ScalarValue representing positive or negative infinity for the same data type:
```

Comment:
👍  I double checked that null represents invite (it is nicely explained in the comments of `Interval`)